src/metrics.jl

```diff
@@ -165,115 +167,117 @@ PeriodicEuclidean() = PeriodicEuclidean(Int[])
 
 ###########################################################
 #
-#  Define Evaluate
+#  Implementations
 #
 ###########################################################
 
 const ArraySlice{T} = SubArray{T,1,Array{T,2},Tuple{Base.Slice{Base.OneTo{Int}},Int},true}
 
 @inline parameters(::UnionMetrics) = nothing
 
-# Specialized for Arrays and avoids a branch on the size
-@inline Base.@propagate_inbounds function evaluate(d::UnionMetrics, a::Union{Array, ArraySlice}, b::Union{Array, ArraySlice})
-    @boundscheck if length(a) != length(b)
-        throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
-    end
-    p = parameters(d)
-    @boundscheck if p !== nothing
-        length(a) != length(p) && throw(DimensionMismatch("arrays have length $(length(a)) but parameters have length $(length(p))."))
-    end
-    if length(a) == 0
-        return zero(result_type(d, a, b))
-    end
-    @inbounds begin
-        s = eval_start(d, a, b)
-        if p === nothing
-            @simd for I in 1:length(a)
-                ai = a[I]
-                bi = b[I]
-                s = eval_reduce(d, s, eval_op(d, ai, bi))
-            end
-        else
-            @simd for I in 1:length(a)
-                aI = a[I]
-                bI = b[I]
-                pI = p[I]
-                s = eval_reduce(d, s, eval_op(d, aI, bI, pI))
-            end
-        end
-        return eval_end(d, s)
-    end
-end
+# breaks the implementation into eval_start, eval_op, eval_reduce and eval_end
+for M in metric_list
```

Comment:
An alternative way is to rename it to `_evaluate` and make this large block into a smaller one.

```julia
for M in metric_list
    @eval @inline (d::$M)(a::AbstractArray, b::AbstractArray) = _evaluate(d, a, b)
    @eval @inline (d::$M)(a::Number, b::Number) = eval_end(d, eval_op(dist, a, b))
```